nomos-services/cryptarchia-consensus/src/lib.rs

```diff
@@ -1366,8 +1410,123 @@ where
+            .prune_old_forks()
+            .collect::<Vec<_>>();
+
+        let Ok(()) = Self::delete_pruned_blocks_from_storage(
+            old_forks_pruned.iter().copied(),
+            storage_adapter,
+        )
+        .await
+        else {
+            // Bail out early (not changing the input `cryptarchia` value if the storage
+            // returned an error).
+            return;
+        };
+
+        // And remove the notes of the pruned blocks from the leader state.
+        for old_fork_pruned in &old_forks_pruned {
+            if local_leader_changes.prune_notes_at(old_fork_pruned) {
+                tracing::debug!(
+                    target: LOG_TARGET,
+                    "Leader proof for block {old_fork_pruned} removed from in-memory storage."
+                );
+            }
+        }
+
+        // If all operations succeed, update the input values to point to the updated
+        // state.
+        *cryptarchia = local_cryptarchia_changes;
+        *leader = local_leader_changes;
+    }
+
+    /// Send a bulk deletion request to the storage adapter.
+    ///
+    /// If no request fails, the method returns `Ok()`.
+    /// If any request fails, the header ID and the generated error are
+    /// collected and returns as part of the `Err` result.
+    async fn delete_pruned_blocks_from_storage<Headers>(
+        pruned_block_headers: Headers,
+        storage_adapter: &StorageAdapter<Storage, TxS::Tx, BS::BlobId, RuntimeServiceId>,
+    ) -> Result<(), Vec<(HeaderId, DynError)>>
+    where
+        Headers: Iterator<Item = HeaderId> + Send,
+    {
+        let blocks_to_delete = pruned_block_headers.collect::<Vec<_>>();
+        let block_deletion_outcomes = blocks_to_delete.iter().copied().zip(
+            storage_adapter
+                .remove_blocks(blocks_to_delete.iter().copied())
+                .await,
+        );
+
+        let errors = block_deletion_outcomes.fold(
+            Vec::with_capacity(blocks_to_delete.len()),
+            |mut errors, (block_id, outcome)| {
+                match outcome {
+                    Ok(Some(_)) => {
+                        tracing::debug!(
+                            target: LOG_TARGET,
+                            "Block {block_id:#?} successfully deleted from storage."
+                        );
+                    }
+                    Ok(None) => {
+                        tracing::trace!(
+                            target: LOG_TARGET,
+                            "Block {block_id:#?} was not found in storage."
+                        );
+                    }
+                    Err(e) => {
+                        tracing::error!(
+                            target: LOG_TARGET,
+                            "Error deleting block {block_id:#?} from storage: {e}."
+                        );
+                        errors.push((block_id, e));
+                    }
+                }
+                errors
+            },
+        );
```

Comment:
a map+collect would be better. I think in this case the compiler already optimizes when it knows the size.

Reply:
I went for a `filter_map`, which felt cleaner: https://github.com/logos-co/nomos/pull/1261/commits/53a56eb1aeff44077e1fe438dbf47cd872ba9dd8. Thanks for the tip tho!